DAS network/validator extension (#2172)
ethereum/eth2.0-specs: specs/phase1/p2p-das.md

#### Horizontal to vertical

Nodes on this same subnet can replicate the sampling efficiently (including a proof for each sample),
and distribute it to any vertical networks that are available to them.

Since the messages are content-addressed (instead of origin-stamped), 
multiple publishers of the same samples on a vertical subnet do not hurt performance, 
but actually improve it by shortcutting regular propagation on the vertical subnet, and thus lowering the latency to a sample.

### Vertical subnets

Vertical subnets propagate the samples to every peer that is interested.
These interests are randomly sampled and rotate quickly: although not perfect, 
sufficient to avoid any significant amount of nodes from being 100% predictable.

As soon as a sample is missing after the expected propagation time window,
nodes can divert to the pull-model, or ultimately flag it as unavailable data.

Note that the vertical subnets are shared between the different shards,
and a simple hash function `(shard, slot, sample_index) -> subnet_index` defines which samples go where.
This is to evenly distribute samples to subnets, even when one shard has more activity than the other.

> TODO: define `(shard, slot, sample_index) -> subnet_index` hash function.

Reviewer: I still think that this is unnecessary complexity, especially with @dankrad's degree change, which means that all blobs now have equal "length", just different "rate", so the workload of each sample index will be identical. So better to just make the subnet index _be_ the sample index.

Author: And how about different powers of two? Or will we always have a constant amount of samples?
And why put samples of different shards on the same subnet, isn't this something we would want to test the performance of?
If we just define a simple function that maps one to the other in a configurable way, we can tune it later.

Reviewer 2: It would always have the same number of samples.

      - [Vertical subnets: `das_sample_{subnet_index}`](#vertical-subnets-das_sample_subnet_index)
  - [DAS in the Req-Resp domain: Pull](#das-in-the-req-resp-domain-pull)
    - [Messages](#messages)
      - [DASQuery](#dasquery)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# Ethereum 2.0 Phase 1 -- Network specification for Data Availability Sampling 

**Notice**: This document is a work-in-progress for researchers and implementers.

## Table of contents

<!-- TOC -->

<!-- /TOC -->

## Introduction

For an introduction about DAS itself, see [the DAS section in the Phase 1 validator spec](./validator.md#data-availability-sampling).
This is not a pre-requisite for the network layer, but will give you valuable context. 

For sampling, all nodes need to query for `k` random samples each slot.

*__TODO__: describe big picture of sampling workload size*

This is a lot of work, and ideally happens at a low latency.

To achieve quick querying, the query model is changed to *push* the samples to listeners instead, using GossipSub.
The listeners then randomly rotate their subscriptions to keep queries unpredictable.

> Except for a small subset of subscriptions, which will function as a backbone to keep topics more stable.

Reviewer 3: ```suggestion
Except for a small subset of subscriptions, which will function as a backbone to keep topics more stable and allow for efficient peer discovery.
```

Publishing can utilize the fan-out functionality in GossipSub, and is easier to split between nodes:
nodes on the horizontal networks can help by producing the same samples and fan-out publishing to their own peers.

This push model also helps to obfuscate the original source of a message:

> the listeners will not have to make individual queries to some identified source.

Reviewer 3: ```suggestion
the listeners do not have to make individual queries to some identified source.
```

The push model does not aim to serve "historical" queries (anything older than the most recent).

> Historical queries are still required for the unhappy case, where messages are not pushed quick enough,

Reviewer 3: Also in the case for historic queries

Author: > historical queries are still required... in case of historical queries?

I think I get what you mean, they are necessary for retrieving older missing data, but we should think of better wording before confusing others more.

and missing samples are not reconstructed by other nodes on the horizontal subnet quick enough.

The main challenge in supporting historical queries is to target the right nodes, 
without concentrating too many requests on a single node, or breaking the network/consensus identity separation.

## DAS Subnets

On a high level, the push-model roles are divided into:
- Sources: create blobs of shard block data, and transformed into many tiny samples. 
- Sinks: continuously look for samples

At full operation, the network has one proposer, per shard, per slot.

In the push-model, there are:
- *Vertical subnets*: Sinks can subscribe to indices of samples: there is a sample to subnet mapping.
- *Horizontal subnets*: Sources need to distribute samples to all vertical networks: they participate in a fanout layer.

### Horizontal subnets

The shift of the distribution responsibility to a proposer can only be achieved with amplification:
a regular proposer cannot reach every vertical subnet.

#### Publishing

> To publish their work, proposers already put the shard block as a whole on a shard-block subnet.

Reviewer 3: ```suggestion
To publish their work, proposers propagate the shard block as a whole on a shard-block subnet.
```

but actually improve it by shortcutting regular propagation on the vertical subnet, and thus lowering the latency to a sample.

### Vertical subnets

Vertical subnets propagate the samples to every peer that is interested.
These interests are randomly sampled and rotate quickly: although not perfect, 
sufficient to avoid any significant amount of nodes from being 100% predictable.

As soon as a sample is missing after the expected propagation time window,
nodes can divert to the pull-model, or ultimately flag it as unavailable data.

Note that the vertical subnets are shared between the different shards,
and a simple hash function `(shard, slot, sample_index) -> subnet_index` defines which samples go where.
This is to evenly distribute samples to subnets, even when one shard has more activity than the other.

TODO: define `(shard, slot, sample_index) -> subnet_index` hash function.

#### Slow rotation: Backbone

To allow for subscriptions to rotate quickly and randomly, a backbone is formed to help onboard peers into other topics.

This backbone is based on a pure function of the *node* identity and time:

> - Nodes can be found *without additional discovery overhead*:

Reviewer 3: Note that we still sometimes search the DHT for peers on these topics

Author: added a note

  peers on a vertical topic can be found by searching the local peerstore for identities that hash to the desired topic(s). 

> - Nodes can be held accountable for contributing to the backbone:

Reviewer 3: Should DAS be a requirement for a node to participate in the network?
I suppose doing a small number, e.g. 4 topics, is maybe a reasonable requirement even for pretty light nodes

Author: Do we have a bandwidth constrain in mind? With a sample of e.g. 256 bytes, 4 topics, with one sample per slot, shouldn't be a problem for even light nodes.

  peers that particpate in DAS but are not active on the appropriate backbone topics can be scored down.

A node should anticipate backbone topics to subscribe to based their own identity.
These subscriptions rotate slowly, and with different offsets per node identity to avoid sudden network-wide rotations.

```python
# TODO hash function: (node, time)->subnets
```

Backbone subscription work is outlined in the [DAS validator spec](./validator.md#data-availability-sampling)

#### Quick Rotation: Sampling

> A node MUST maintain `k` random subscriptions to topics, and rotate these according to the [DAS validator spec](./validator.md#data-availability-sampling).

Reviewer 3: If a _node_ (rather than just a _validator_ node) must maintain these topics and rotate them, then the rotation and sampling procedure maybe shouldn't go in the validator guide and instead in a free-standing DAS guide.

Author: We now have 3 docs:
- das-internals: the commitments, proofs, sampling, etc.
- das-p2p: about the network layer
- das-participation: about operating DAS

As well as a new phase1 network doc to cover the general things (shard headers and blobs)

> If the node does not already have connected peers on the topic it needs to sample, it can search its peerstore for peers in the topic backbone.

Reviewer 3: ```suggestion
If the node does not already have connected peers on the topic it needs to sample, it can first search its peerstore and, if necessary, in the DHTfor peers in the topic backbone.
```

If participating in DAS, upon receiving a `signed_blob` for the first time, with a `slot` not older than `MAX_RESAMPLE_TIME`,
a subscriber of a `shard_blob_{shard}` SHOULD reconstruct the samples and publish them to vertical subnets.
Take `blob = signed_blob.blob`:
1. Extend the data: `extended_data = extend_data(blob.data)`
2. Create samples with proofs: `samples = sample_data(blob.slot, blob.shard, extended_data)`
3. Fanout-publish the samples to the vertical subnets of its peers (not all vertical subnets may be reached).

The [DAS validator spec](./validator.md#data-availability-sampling) outlines when and where to participate in DAS on horizontal subnets.

The following validations MUST pass before forwarding the `signed_blob` (with inner `blob`) on the horizontal subnet or creating samples for it.
- _[REJECT]_ `blob.shard` MUST match the topic `{shard}` parameter. (And thus within valid shard index range)
- _[IGNORE]_ The `blob` is not from a future slot (with a `MAXIMUM_GOSSIP_CLOCK_DISPARITY` allowance) --
  i.e. validate that `blob.slot <= current_slot`
  (a client MAY queue future blobs for processing at the appropriate slot).

> - _[IGNORE]_ The blob is the first blob with valid signature received for the proposer for the `(slot, shard)`: `blob.slot`.

Reviewer 3: what do you mean by "`(slot, shard)`: `blob.slot`"?

Author: Think I intended to include the signature validation, but it doesn't read right indeed. Will fix, thanks

> - _[REJECT]_ As already limited by the SSZ list-limit, it is important the blob is well-formatted and not too large.

Reviewer 3: This just means the SSZ typing is correct and it can be decoded?

Author: Yes. Thought I would make it extra explicit, since it's a SSZ type not seen on the consensus layer, and clients may only be implementing the serialization/deserialization part.

- _[REJECT]_ The `blob.data` MUST NOT contain any point `p >= MODULUS`. Although it is a `uint256`, not the full 256 bit range is valid.
- _[REJECT]_ The proposer signature, `signed_blob.signature`, is valid with respect to the `proposer_index` pubkey, signed over the SSZ output of `commit_to_data(blob.data)`.

> - _[REJECT]_ The blob is proposed by the expected `proposer_index` for the blob's slot

Reviewer 3: ```suggestion
- _[REJECT]_ The blob is proposed by the expected `proposer_index` for the blob's slot.
```

TODO: make double blob proposals slashable?

#### Vertical subnets: `das_sample_{subnet_index}`

Shard blob samples can be verified with just a 48 byte KZG proof (commitment quotient polynomial), 
against the commitment to blob polynomial, specific to that `(shard, slot)` key.

The following validations MUST pass before forwarding the `sample` on the vertical subnet.

> - _[IGNORE]_ The commitment for the (`sample.shard`, `sample.slot`, `sample.index`) tuple must be known.

Reviewer 3: This comes from the global `ShardHeader` topic?

Author: We do not have a general phase1 network doc yet, but we do have a das network doc now. I listed shard-headers for completeness. And it will likely involve some validator spec in how to handle these headers.

From the PR description: "Time for a general phase1 network doc, to cover propagation of shard headers and such?" - That's a yes then. Include it in this same PR?

Author: Split out the header topic (and fixed it to be a global topic), see new phase1 network doc.

    - `sample_count = (points_count + POINTS_PER_SAMPLE - 1) // POINTS_PER_SAMPLE`
    - `points_count` is the length as claimed along with the commitment, which must be smaller than `MAX_SAMPLES_PER_BLOCK`.
- _[IGNORE]_ The `sample` is not from a future slot (with a `MAXIMUM_GOSSIP_CLOCK_DISPARITY` allowance) --
  i.e. validate that `sample.slot <= current_slot`. A client MAY queue future samples for processing at the appropriate slot, if it passed formatting conditions.
- _[IGNORE]_ This is the first received sample with the (`sample.shard`, `sample.slot`, `sample.index`) key tuple.
- _[REJECT]_ As already limited by the SSZ list-limit, it is important the sample data is well-formatted and not too large.
- _[REJECT]_ The `sample.data` MUST NOT contain any point `p >= MODULUS`. Although it is a `uint256`, not the full 256 bit range is valid.
- _[REJECT]_ The `sample.proof` MUST be valid: `verify_sample(sample, sample_count, commitment)`

Upon receiving a valid sample, it SHOULD be retained for a buffer period, if the local node is part of the backbone that covers this sample.
This is to serve other peers that may have missed it.

> ## DAS in the Req-Resp domain: Pull

Author: With the gossip backbone, we already have some properties similar to a DHT, but for a short amount of time (no big incentives for this work, but also not so costly to buffer a subset of samples for a short period of time):
- hashing content to network nodes, without need of discovery overhead (since it's based on node identity)
- distributing content to them (the initial gossip behavior, or they too do similar requests if they miss also)
- multiple nodes, redundancy
- randomly assigned
- expectation that they get to see the data.

And gossipsub is more like a blackbox library sometimes, we have parameters to tweak, but no direct control over IWANT/IHAVE. So to cover short-term catchup (e.g. after a node restart, or after some gossip problem), it's nice to catch up using infrastructure that's almost entirely there already (just need to define buffering, but backbone is there for gossipsub anyway).

Then agree, for as long as the custody period, we have to look at better solutions. Probably a DHT of some kind, and incorporate the custody functionality properly.
